### PR TITLE
EvaluateInputTensor callback to support validation with input tensors

### PR DESCRIPTION
**Pull request by ahundt**
There isn't currently a way to perform validation after each epoch with input tensors when both `Input(tensor=tensor)` and `Model.compile(target_tensors=tensors)` are both defined.
This adds a very simple callback which accepts a validation model while specifically minimizing changes to Keras' training internals.

Here is the expected output of `mnist_tfrecord.py`:
```
Extracting MNIST-data/train-images-idx3-ubyte.gz
Extracting MNIST-data/train-labels-idx1-ubyte.gz
Extracting MNIST-data/t10k-images-idx3-ubyte.gz
Extracting MNIST-data/t10k-labels-idx1-ubyte.gz
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (100, 28, 28, 1)          0         
_________________________________________________________________
conv2d_1 (Conv2D)            (100, 26, 26, 32)         320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (100, 13, 13, 32)         0         
_________________________________________________________________
conv2d_2 (Conv2D)            (100, 11, 11, 64)         18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (100, 5, 5, 64)           0         
_________________________________________________________________
flatten_1 (Flatten)          (100, 1600)               0         
_________________________________________________________________
dense_1 (Dense)              (100, 512)                819712    
_________________________________________________________________
dropout_1 (Dropout)          (100, 512)                0         
_________________________________________________________________
x_train_out (Dense)          (100, 10)                 5130      
=================================================================
Total params: 843,658
Trainable params: 843,658
Non-trainable params: 0
_________________________________________________________________
Epoch 1/5
100/100 [==============================] - 8s 80ms/stepss: 0.1560 - acc: 0.9517      
val_loss: 0.04948705253300432 val_acc: 0.985000005364418 
600/600 [==============================] - 95s 158ms/step - loss: 0.1557 - acc: 0.9518
Epoch 2/5
100/100 [==============================] - 1s 6ms/steposs: 0.0505 - acc: 0.9843 
val_loss: 0.03512724666628855 val_acc: 0.9885000056028366 
600/600 [==============================] - 11s 19ms/step - loss: 0.0503 - acc: 0.9844
Epoch 3/5
100/100 [==============================] - 1s 5ms/steposs: 0.0385 - acc: 0.9884 
val_loss: 0.030810652751943054 val_acc: 0.9896000051498413 
600/600 [==============================] - 11s 18ms/step - loss: 0.0386 - acc: 0.9884
Epoch 4/5
100/100 [==============================] - 1s 5ms/steposs: 0.0304 - acc: 0.9907 
val_loss: 0.034214809939185215 val_acc: 0.9891000074148179 
600/600 [==============================] - 11s 18ms/step - loss: 0.0305 - acc: 0.9907
Epoch 5/5
100/100 [==============================] - 1s 5ms/steposs: 0.0265 - acc: 0.9918
val_loss: 0.030512848425833 val_acc: 0.9911000055074691 
600/600 [==============================] - 10s 17ms/step - loss: 0.0265 - acc: 0.9918
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 1600)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               819712    
_________________________________________________________________
dropout_1 (Dropout)          (None, 512)               0         
_________________________________________________________________
x_train_out (Dense)          (None, 10)                5130      
=================================================================
Total params: 843,658
Trainable params: 843,658
Non-trainable params: 0
_________________________________________________________________
5000/5000 [==============================] - 0s 65us/step

Test accuracy: 0.990400009155

```

credit to @DingYu95 for the initial implementation of this callback, which I simplified for this PR.

### CODE REVIEW

**Comment by fchollet** [comment]
The current plan is to *eventually* support data tensors in `fit`/`predict`/etc, including for validation, by using TF graph rewiring. There's no timeline, but I am confident it will happen in the near future.

As such, such a callback could not become a permanent part of the Keras API. So I suggest putting the callback code in the example script directly. That makes the example script a bit long, but I think that's fine.

**Comment by ahundt** [comment]
Moved everything into mnist_tfrecord and dropped the unit test.

There are *lots of little things broken with input tensors*:

- Partial Callback support (example: TensorBoard shows no images or gradients)
- You can't get all your tensors back from one backend `session.run()` call, as per https://github.com/keras-team/keras/pull/9121.
- Many of the nice compile checks with clear feedback because the `model.fit(None, None)` hacks are needed.

Probably the best supported Keras + `Input(tensor=op)` interface is this [tfrecord generator hack](https://github.com/tensorflow/tensorflow/issues/8787#issuecomment-357289279):
 
```
def` tensors2gen(*input_tensors):
    while True:
        data = sess.run(input_tensors)
        if len(input_tensors) == 1:
            data = data[0]
        yield data
```

Rather than thinking of all these as tf only functionality aren't changes like this and #9121 a way of simply improving support for the tf backend without taking away from the others?

**Comment by ahundt** [comment]
I don't think the travis failure is due to this PR.
